### PR TITLE
extract julia directly to tool path to maintain mtimes

### PR DESCRIPTION
**Pull request by IanButterworth**
A bit of a hack to get julia installed with original mtimes to help https://github.com/JuliaLang/julia/issues/50667

The comments explain the reasoning.

Also should be a bit faster as it avoids the `cp` action.

You can see here that 1.10.0 has the release mtimes
```
StatStruct("/Users/runner/hostedtoolcache/julia/1.10.0/x64/share/julia/stdlib/v1.10/MbedTLS_jll/src/MbedTLS_jll.jl" 
size: 2169 bytes device: 16777220 inode: 7005911 mode: 0o100644 (-rw-r--r--) nlink: 1 uid: 501 (runner) 
gid: 20 (staff) rdev: 0 blksz: 4096 blocks: 8 
mtime: 2023-12-25T20:50:13+0000 (9 days ago) ctime: 2024-01-04T06:25:27+0000 (8 seconds ago))
```

The added mtime test fails on master to illustrate the issue. https://github.com/julia-actions/setup-julia/pull/200

Closes https://github.com/julia-actions/setup-julia/pull/100
Closes https://github.com/julia-actions/setup-julia/pull/165

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Can we add a test for this? E.g. do `StatStruct` in Julia (like in your example above) and check the mtime?

File: src/setup-julia.ts [hunk 1]
```diff
@@ -69,14 +69,20 @@ async function run() {
 
         if (!juliaPath) {
             core.debug(`could not find Julia ${arch}/${version} in cache`)
-            const juliaInstallationPath = await installer.installJulia(versionInfo, version, arch)
 
-            // Add it to cache
-            juliaPath = await tc.cacheDir(juliaInstallationPath, 'julia', version, arch)
+            // we want julia to be installed with unmodified file mtimes
```
Comment (DilumAluthge) [inline src/setup-julia.ts:73]:
```suggestion
            // https://github.com/julia-actions/setup-julia/pull/196
            // we want julia to be installed with unmodified file mtimes
```

**Comment by IanButterworth** [comment]
Added mtime test which fails on master https://github.com/julia-actions/setup-julia/pull/200

File: .github/workflows/example-builds-defaultarch.yml [hunk 2]
```diff
@@ -44,3 +44,17 @@ jobs:
           version: ${{ matrix.julia-version }}
       - run: julia --version
       - run: julia --compile=min -O0 -e 'import InteractiveUtils; InteractiveUtils.versioninfo()'
+      - name: "Check that the correct julia is used and that archive mtimes are maintained"
+        run: |
+          if !occursin("hostedtoolcache", Sys.BINDIR)
+            error("the wrong julia is being used: $(Sys.BINDIR)")
+          end
+          if VERSION > v"1.7.0-0" # pkgdir was introduced here, and before then mtime wasn't a problem so just skip
```
Comment (omus) [inline .github/workflows/example-builds-defaultarch.yml:52]: Based upon the comment it sounds like this version includes `pkgdir`:
```suggestion
          if VERSION >= v"1.7.0-0" # pkgdir was introduced here, and before then mtime wasn't a problem so just skip
```

File: .github/workflows/example-builds-nightly-defaultarch.yml [hunk 3]
```diff
@@ -42,3 +42,17 @@ jobs:
...
+          end
+          if VERSION > v"1.7.0-0" # pkgdir was introduced here, and before then mtime wasn't a problem so just skip
+            using Pkg
+            src = pkgdir(Pkg, "src", "Pkg.jl")
+            # mtime is when it's compressed, ctime is when the file is extracted
+            if !<(mtime(src), ctime(src))
+              error("source mtime ($(mtime(src))) is not earlier than ctime ($(ctime(src)))")
+            end
+          end
+        shell: julia --startup-file=no --color=yes {0}
```
Comment (omus) [inline .github/workflows/example-builds-nightly-defaultarch.yml:58]: You can avoid the repeating this step in multiple workflows by using another job with the [`uses` field](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_iduses).

File: src/setup-julia.ts [hunk 4]
```diff
@@ -69,14 +69,21 @@ async function run() {
...
-            juliaPath = await tc.cacheDir(juliaInstallationPath, 'julia', version, arch)
+            // https://github.com/julia-actions/setup-julia/pull/196
+            // we want julia to be installed with unmodified file mtimes
+            // but tc.cacheDir uses `cp` which destroys mtime
+            // and `tc` provides no API to get the tool directory alone
+            // so hack it by installing a dummy julia file then use the path it returns
+            // and extract the archives directly to that location
+            const tempDummyDir = fs.mkdtempSync('julia-dummy-')
+            juliaPath = await tc.cacheDir(tempDummyDir, 'julia', version, arch)
+            await installer.installJulia(juliaPath, versionInfo, version, arch)
```
Comment (omus) [inline src/setup-julia.ts:81]: Should be equivalent but uses the more descriptive temporary directory name:
```suggestion
            const tempDir = fs.mkdtempSync(`julia-${arch}-${version}-`)
            juliaInstallDir = await tc.cacheDir(tempDir, 'julia', version, arch)
            await installer.installJulia(juliaInstallDir, versionInfo, version, arch)
```

File: .github/workflows/example-builds-defaultarch.yml (hunk 2 above)
Comment (DilumAluthge) [reply .github/workflows/example-builds-defaultarch.yml:52]: FWIW, you can make this a bit more concise: `if VERSION >= v"1.7.0-"`

File: src/setup-julia.ts (hunk 4 above)
Comment (IanButterworth) [reply src/setup-julia.ts:81]: Naming it like that makes it seem like it has to be a special name. It doesn't, just needs to be a temp dir.

As for renaming `juliaPath` that won't work as it's a variable defined and used outside of this scope

File: .github/workflows/example-builds-nightly-defaultarch.yml (hunk 3 above)
Comment (IanButterworth) [reply .github/workflows/example-builds-nightly-defaultarch.yml]: Having a common .jl file is effective enough for this I think

File: src/setup-julia.ts (hunk 4 above)
Comment (omus) [reply src/setup-julia.ts:81]: Upon further reflection I agree with your points. I do think we should just name this `emptyDir` to make it clearer that this directory is empty and will always be empty:
```suggestion
            const emptyDir = fs.mkdtempSync('empty')
            juliaPath = await tc.cacheDir(emptyDir, 'julia', version, arch)
            await installer.installJulia(juliaPath, versionInfo, version, arch)
```

File: .github/workflows/common-tests.jl [hunk 5]
```diff
@@ -0,0 +1,11 @@
+if !occursin("hostedtoolcache", Sys.BINDIR)
+    error("the wrong julia is being used: $(Sys.BINDIR)")
+end
+if VERSION >= v"1.7.0" # pkgdir was introduced here, and before then mtime wasn't a problem so just skip
+    using Pkg
+    src = pkgdir(Pkg, "src", "Pkg.jl")
+    # mtime is when it's compressed, ctime is when the file is extracted
+    if !<(mtime(src), ctime(src))
```
Comment (omus) [inline .github/workflows/common-tests.jl:8]: Any reason you chose this over the more readable:
```suggestion
    if mtime(src) >= ctime(src)
```

File: .github/workflows/common-tests.jl [hunk 6]
```diff
@@ -0,0 +1,11 @@
+if !occursin("hostedtoolcache", Sys.BINDIR)
```
Comment (omus) [inline .github/workflows/common-tests.jl:1]: I'd consider moving this into a `.github/bin` or `.github/scripts` directory as this file isn't a GitHub workflow

File: src/setup-julia.ts [hunk 7]
```diff
@@ -69,14 +69,21 @@ async function run() {
...
-            const juliaInstallationPath = await installer.installJulia(versionInfo, version, arch)
 
-            // Add it to cache
-            juliaPath = await tc.cacheDir(juliaInstallationPath, 'julia', version, arch)
+            // https://github.com/julia-actions/setup-julia/pull/196
+            // we want julia to be installed with unmodified file mtimes
+            // but tc.cacheDir uses `cp` which destroys mtime
+            // and `tc` provides no API to get the tool directory alone
+            // so hack it by installing a dummy julia file then use the path it returns
+            // and extract the archives directly to that location
```
Comment (omus) [inline src/setup-julia.ts:78]:
```suggestion
            // we want julia to be installed with unmodified file mtimes
            // but `tc.cacheDir` uses `cp` internally which destroys mtime
            // and `tc` provides no API to get the tool directory alone
            // so hack it by installing a empty directory then use the path it returns
            // and extract the archives directly to that location
```

**Review by omus — APPROVED** [review]
Changes are reasonable to avoid having the `tool-cache` module from modifying the mtimes. My only concern is that the `tc.cacheDir` function is described as ["Caches a directory and installs it to the tool cacheDir"](https://github.com/actions/toolkit/blob/5430c5d84832076372990c7c27f900878ff66dc9/packages/tool-cache/src/tool-cache.ts#L410) so it is possible that a non-breaking change to `tc.cacheDir` such as making the tool path immutable would break this code. However, since the version of `@actions/tool-cache` is pinned until we update it we should be able to avoid such a scenario.

File: .github/workflows/common-tests.jl (hunk 5 above)
Comment (IanButterworth) [reply .github/workflows/common-tests.jl:8]: Because it matched the error message

Comment (IanButterworth) [reply .github/workflows/common-tests.jl:8]: But I'll go for your suggestion for code readabilty

File: lib/setup-julia.js [hunk 8]
```diff
@@ -93,16 +93,16 @@ function run() {
...
-                const tempDummyDir = fs.mkdtempSync('julia-dummy-');
-                juliaPath = yield tc.cacheDir(tempDummyDir, 'julia', version, arch);
+                const emptyDir = fs.mkdtempSync('empty');
+                juliaPath = yield tc.cacheDir(emptyDir, 'julia', version, arch);
                 yield installer.installJulia(juliaPath, versionInfo, version, arch);
                 core.debug(`added Julia to cache: ${juliaPath}`);
-                // Remove temporary dummy dir
-                fs.rmSync(tempDummyDir, { recursive: true });
+                // Remove empty dir
+                fs.rmSync(emptyDir);
```
Comment (omus) [inline lib/setup-julia.js:105]: I think we want [`rmdirSync`](https://nodejs.org/api/fs.html#fsrmdirsyncpath-options) now: 
```suggestion
                fs.rmdirSync(emptyDir);
```

**Review by omus — APPROVED** [review]
:chefkiss:

**Comment by DilumAluthge** [comment]
Can we remove the tests from here, and keep the tests in #200, and merge #200 separately?

Reason: I am hoping that this PR will be a short-term solution; I'm hoping that the long-term solution will be for me to get a patch into upstream `tc.cacheDir` to add the ability to preserve mtimes. If we are able to get that feature accepted into upstream, then we can revert this PR. But if/when we revert this PR, I wouldn't want to revert the tests. If we keep the tests separate in #200, then later we can revert this PR without reverting #200.

**Comment by omus** [comment]
> I'm hoping that the long-term solution will be for me to get a patch into upstream `tc.cacheDir` to add the ability to preserve mtimes.

If we're changing `tool-cache` you may want to introduce a `tc.cache` which provides a callback so we just do the install inside the callback. I foresee that this new function could be used by `tc.cacheFile` and `tc.cacheDir`.

**Comment by IanButterworth** [comment]
Ok, #200 is updated. Probably best to merge that as failing first then rebase this

**Review by DilumAluthge — APPROVED** [review]

**Comment by DilumAluthge** [comment]
This is great!

Is this the last change needed to make `/compiled` reusable in GitHub Actions CI when using Julia 1.10?

**Comment by IanButterworth** [comment]
Yeah, should be.

**Comment by IanButterworth** [comment]
I make a test release for 1.9.5 using `bin/build-test-release.sh` and tested it here https://github.com/IanButterworth/PowerMonitor.jl/pull/2

with
```
uses: julia-actions/setup-julia@test/ib/1_9_5/releases/v1.9.5
```

And all platforms avoid re-precompilation once cached

**Comment by IanButterworth** [comment]
Released now https://github.com/julia-actions/setup-julia/releases/tag/v1.9.5

**Comment by omus** [comment]
> Can we remove the tests from here, and keep the tests in #200, and merge #200 separately?
> 
> Reason: I am hoping that this PR will be a short-term solution; I'm hoping that the long-term solution will be for me to get a patch into upstream `tc.cacheDir` to add the ability to preserve mtimes. If we are able to get that feature accepted into upstream, then we can revert this PR. But if/when we revert this PR, I wouldn't want to revert the tests. If we keep the tests separate in #200, then later we can revert this PR without reverting #200.

Upstream `tc.cacheDir` PR to preserve mtimes: https://github.com/actions/toolkit/pull/1617

**Pull request merged**